===== Patch instructions generated by SAVER =====
[+] { Insert: if ({ proxy_C_OpenSession:n$18 == { 0 } }) free(*(proxy_C_OpenSession:sess)) at 16 (line 683, column 5) }
===== Actual patch in diff format =====
@@ -678,9 +678,12 @@ proxy_C_OpenSession (CK_X_FUNCTION_LIST *self,
 				sess->wrap_slot = map.wrap_slot;
 				sess->real_session = *handle;
 				sess->wrap_session = ++state->last_handle; /* TODO: Handle wrapping, and then collisions */
-				if (!p11_dict_set (state->px->sessions, &sess->wrap_session, sess))
+				ret = p11_dict_set (state->px->sessions, &sess->wrap_session, sess);
+				if (!ret)
 					warn_if_reached ();
 				*handle = sess->wrap_session;
+				if (ret == 0)
+					free(sess);
 			}
 
 		p11_unlock ();
